### PR TITLE
[pro#590] Rake task to create a new webhook endpoint

### PR DESCRIPTION
**Pull request by lizconlan**
## Relevant issue(s)

Closes mysociety/alaveteli-professional#590

## What does this do?

Adds a rake task to:
  * Create a new Stripe webhook endpoint with the right:
    * url
    * api_version
    * list of subscribed events

From 2428a5d on it also:
  * Introduces a simple logging mechanism to record which webhooks we have already processed
  * Detects already-processed (duplicate) webhooks
    * Avoids processing them twice
    * Responds with a custom 200 error so they do not keep getting resent
  * Tidies up (disables) duplicate webhook endpoints after the new endpoint has been created to avoid having live duplicate endpoints listening for events and sending them to the app

## Why was this needed?

This solves a couple of problems for us:

We can set the Stripe API version for the webhooks so - although the wrapped Event data will still use the format set by the Stripe account (new accounts will take whatever the latest version is at the time they are created) - the webhook wrapper will take the expected form as set by the api_version.

Once the site re-users have set up a Stripe account and shared the API keys with us, we can create - and potentially maintain - the webhook endpoint without asking them to do it via the Stripe admin tools. We - quite rightly - can't see their Stripe settings so a mistake here would be quite hard to find and debug.

## Implementation notes

It currently disables rather than deleting the out of date endpoints because we've currently got the Stripe gem locked to v3.29.0 and the ability to use the `delete` method of the APIs is not introduced until [v4.12.0](https://github.com/stripe/stripe-ruby/releases/tag/v4.12.0)

This is not ideal as it leaves stray endpoints in the account - with Stripe in test mode this can result in an error when trying to add a 4th new endpoint with a different API version, not sure if we'd encounter the same problem under live conditions - rather than removing them. 

## Notes to reviewer

Presently the cleanup portion of the script does not check whether the endpoint it already disabled and sends `update(disabled: true)` regardless, which does no harm but is a bit wasteful. Not sure whether to note the problem and leave it for now, add a check for the endpoint being disabled, or start compatibility testing against  [Stripe v4.12.0](https://github.com/stripe/stripe-ruby/releases/tag/v4.12.0) to see if we would need to modify our code to use it.

Commits prior to https://github.com/mysociety/alaveteli/pull/5302/commits/2942119c73c1fb6497039d792d890b7ccae9abc9 have been merged in from open PRs https://github.com/mysociety/alaveteli/pull/5298 and https://github.com/mysociety/alaveteli/pull/5299 which set the Stripe api_version and upgrade the Stripe gem to v3.29.0 so that we can use the `WebhookEndpoints` API

### CODE REVIEW

File: lib/tasks/stripe.rake [hunk 1]
```diff
@@ -0,0 +1,37 @@
+namespace :stripe do
+  desc "Create the Stripe webhook endpoint"
+  task create_webhook_endpoint: :environment do
+    endpoint = AlaveteliPro::WebhookEndpoints.webhook_endpoint_url
+
+    # Find all hooks that POST to our Alaveteli install
+    hooks = AlaveteliPro::WebhookEndpoints.retrieve_all_endpoints_data.
+              select { |hook| hook.url == endpoint }
+
+    matching_hook = hooks.find { |hook| hook.api_version == Stripe.api_version }
```
[the remark this answers is not in the document]
Comment (lizconlan) [reply lib/tasks/stripe.rake]: Hmm, to me `find` is easier to understand at a glance but the argument for the rule is that `Array#detect` and `Array#find` are the same but that `.find` could be mistaken for the `ActiveRecord` method 🤷‍♀

File: app/controllers/alaveteli_pro/stripe_webhooks_controller.rb [hunk 2]
```diff
@@ -100,6 +115,12 @@ def check_for_event_type
     end
   end
 
+  def check_for_duplicate_event
```
[the remark this answers is not in the document]
Comment (garethrees) [reply app/controllers/alaveteli_pro/stripe_webhooks_controller.rb]: Yeah I think we decided not to worry about this for now as we weren't going to handle upgrades in API version at this time. I think this adds a lot of overlap with the `Webhook` class, so I think we'd want to get a better sense of the overall design before we introduced this

File: spec/lib/alaveteli_pro/webhook_endpoints_spec.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  describe '.retrieve_endpoints' do
+    subject { described_class.retrieve_endpoints }
+
+    let!(:endpoints) do
+      2.times { StripeMock.mock_webhook_event('customer.subscription.deleted') }
+    end
+
+    # StripeMock does not currently support the WebhookEndpoint endpoint
+    it 'should return the expected number of endpoints'
```
[the remark this answers is not in the document]
Comment (lizconlan) [reply spec/lib/alaveteli_pro/webhook_endpoints_spec.rb]: Left pending on purpose as StripeMock doesn't support that part of the API

**Comment by lizconlan** [comment]
I'm going to close this in favour of a simpler one. The later commits as flagged up in the initial PR comment were intended to be a spike/draft/idea of how it could work, but I did not communicate that well.

File: lib/tasks/stripe.rake [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      $stderr.puts 'Webhook endpoint already exists, continuing to cleanup'
+    else
+      # If there's no hook that matches our requirements, create it.
+      Stripe::WebhookEndpoint.create(url: endpoint,
+                                     api_version: Stripe.api_version,
+                                     enabled_events: [
+                                       'customer.subscription.deleted',
+                                       'invoice.payment_succeeded',
+                                       'invoice.payment_failed',
+                                       'customer.subscription.updated'
```
[the remark this answers is not in the document]
Comment (lizconlan) [reply lib/tasks/stripe.rake]: That's a good idea - thanks!

File: lib/alaveteli_pro/webhook_endpoints.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
+module AlaveteliPro
+  ##
+  # A class for working with the Stripe WebhookEndpoints API
+  class WebhookEndpoints
+    ##
+    # Calculates the URL of the Stripe webhook endpoint for the install
+    def self.webhook_endpoint_url
+      protocol = AlaveteliConfiguration.force_ssl ? 'https' : 'http'
+      endpoint = "#{ protocol }://#{ AlaveteliConfiguration.domain }/" \
+                 "pro/subscriptions/stripe-webhook"
```
[the remark this answers is not in the document]
Comment (lizconlan) [reply lib/alaveteli_pro/webhook_endpoints.rb]: Hmm... good point - that made more sense when it was literally in a rake file

**Comment by lizconlan** [comment]
Closed in favour of #5315

File: lib/alaveteli_pro/webhook_endpoints.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (lizconlan) [reply lib/alaveteli_pro/webhook_endpoints.rb]: Oh, that's harder than it looks as just using the `url_helpers` from the module seems to bypass the routing filter and the http/https check

In other words we can end up with:
```diff
- https://test.host/pro/subscriptions/stripe-webhook
+ http://test.host/en/pro/subscriptions/stripe-webhook
```

🤔

[the remark this answers is not in the document]
Comment (gbp) [reply lib/alaveteli_pro/webhook_endpoints.rb]: I think you can skip this by passing `locale: false` into the url helper

[the remark this answers is not in the document]
Comment (lizconlan) [reply lib/alaveteli_pro/webhook_endpoints.rb]: Yep, that fixed it!

[the remark this answers is not in the document]
Comment (lizconlan) [reply lib/alaveteli_pro/webhook_endpoints.rb]: Oh, turns out `include_default_locale_in_urls` defaults to true. Fun times

**Pull request closed without merging**